boards: arc: add initial support for ARC axs board.

IRISZZW opened the pull request:
add basic support for  GPIO and UART with no
interrupt, in axs board, IPs in mother board use
a second-level interrupt which currently not supported.
currently, we can use the debug uart and gpio buttons,
leds and seven-segment deisplays.

dbkinder commented on boards/arc/axs/doc/index.rst:
```diff
@@ -0,0 +1,206 @@
+Overview
+********
+
+The DesignWare(R) ARC(R) Software Development Platforms are complete, standalone
```
Let's simplify that first sentence with:
```
The DesignWare(R) ARC(R) Software Development Platforms
are a mainboard with a CPU card that provide complete, standalone
platforms for software development, debugging, and profiling.
```

dbkinder commented on boards/arc/axs/doc/index.rst:
```diff
@@ -0,0 +1,206 @@
+Hardware
+********
+
+For hardware feature details, refer to :
```
This would look better as a list:
```
For hardware feature details, refer to:

- `DesignWare ARC AXC001 CPU Card Datasheet
  <https://www.synopsys.com/dw/doc.php/ds/cc/arc_axc001_cpu_card.pdf>`_
- `DesignWare ARC AXC003 CPU Card Datasheet
  <https://www.synopsys.com/dw/doc.php/ds/cc/arc_axc003_cpu_card.pdf>`_
- `DesignWare ARC AXS101 Software Development Platform Datasheet
  <https://www.synopsys.com/dw/doc.php/ds/cc/arc_axs101_sdp.pdf>`_
- `DesignWare ARC AXS103 Software Development Platform Datasheet
  <https://www.synopsys.com/dw/doc.php/ds/cc/arc_axs103_sdp.pdf>`_
```

dbkinder commented on boards/arc/axs/doc/index.rst:
```diff
@@ -0,0 +1,206 @@
+To run Zephyr application on AXS Software Development platform,
+you need to setup the board correctly.
+
+* Connect the digilent usb cable from your host to the board.
```
capitalize ``USB``

dbkinder commented on boards/arc/axs/doc/index.rst:
```diff
@@ -0,0 +1,206 @@
+   :goals: debug
+
+At this point you can do your normal debug session. Set breakpoints and then
+'c' to continue into the program.
```
change ``'c'`` to ``:kbd:`c` ``

dbkinder commented on boards/arc/axs/doc/index.rst:
```diff
@@ -0,0 +1,206 @@
+   :alt: DesignWare(R) ARC(R) AXS Software Development platform (synopsys.com)
+
+For details about the board, see: `DesignWare ARC Software Development Platforms
+<https://www.synopsys.com/dw/ipdir.php?ds=arc-software-development-platform>`__
```
Use a single underscore for the outbound link (instead of two underscores).

dbkinder commented on boards/arc/axs/doc/index.rst:
```diff
@@ -0,0 +1,206 @@
+it using openocd and gdb. The program can be download via the USB cable into
+the code and data memories.
+
+Release Notes
```
If there's no content in this section, let's delete this "Release Notes" heading (and underline)

dbkinder commented on boards/arc/axs/doc/index.rst:
```diff
@@ -0,0 +1,206 @@
+
+* Connect the 12V DC power supply to your board
+
+Set up Zephyr Software
```
If there's no content for this section, then delete the "Set up Zephyr Software" heading (and underline)

dbkinder reviewed (approved):
+1 for doc changes, thanks!

vonhust reviewed (approved):
we need to get this merged in 2.2 release

vonhust commented:
@ruuddw @abrodkin  we need to get this merged in release 2.2. The CI failure seems not caused by this  PR.

ruuddw commented on boards/arc/axs/support/openocd.cfg:
```diff
@@ -0,0 +1,73 @@
+#
+# HS Development Kit SoC.
+#
+# Contains quad-core ARC HS38.
```
AXS 103 is not really a quad core. As far as I know, it can be configured as either a HS38 dual core, or as a single HS36/34.
Current initial support is for HS38 dual core mode, right?

IRISZZW replied:
currently, SMP on axs103 not supported. I will clarify it.

ruuddw replied:
It is still a bit confusing and inconsistent. This openocd config contains multiple CPU references/names: EM, HS38, L2 cache, etc. while the dts file has a single hs48.
Please make consistent. If for the openocd file the specific processor is a don't care, then use some generic name, like HS3x/4x, or simply ARC HS or so.

ruuddw commented on boards/arc/axs/doc/index.rst:
```diff
@@ -0,0 +1,197 @@
+The DesignWare(R) ARC(R) Software Development Platforms are a mainboard
+with a CPU card that provide complete, standalone platforms for software
+development, debugging, and profiling. Synopsys currently has two Software
+Development Platforms available: AXS101 and AXS103
```
Do we plan Zephyr support for AXS101? Or would the support here be for AXS103 only?
Maybe use axs103 as name to make clear e.g. the ARC600/700 on axs101 are not planned to be supported.

IRISZZW replied:
currently, we only support AXS103, I will clarify it.

abrodkin replied:
I guess AXS101 makes no sense here because it is based on ARCompact cores (ARC700 & ARC600) which we don't support in Zephyr. So indeed AXS103 might be a better name for the Zephyr board.

ruuddw commented on dts/arc/arc_axs.dtsi:
```diff
@@ -0,0 +1,76 @@
+
+		cpu@0 {
+			device_type = "cpu";
+			compatible = "snps,archs48";
```
AXS103 has HS3x not HS4x

IRISZZW replied:
AXS103 1.3 version support HS4x.

abrodkin replied:
Well in fact we have 2 flavors of AXS103:
1. FPGA firmware v1.2: it has ARC HS38x2
2. FPGA firmware v1.3: it has ARC HS48x2 or ARC HS47D (depending on used .bit-file)

But if we don't need DSP extensions we may treat both v1.2 with HS38x2 & v1.3 with HS48x2 as the same core because from software perspective they are very similar. And if we need to do something specific per core flavor we may do so in run-time based on read IDENTITY register.

Though given we're discussing very optimized RTOS we may still prefer to have 2 a bit separate boards like "axs103_hs38x2" & "axs103_hs48x2" because actually AXS103 v1.2 has also ARC HS36 in FPGA configuration (and is selectable via dip-switch on the board) so there might be 3rd "Zephyr board" like "axs103_hs36" (note it's a single-core configuration).

ruuddw commented on dts/arc/arc_axs.dtsi:
```diff
@@ -0,0 +1,76 @@
+
+#include "skeleton.dtsi"
+
+#include <dt-bindings/i2c/i2c.h>
```
i2c isn't supported yet, right? why include?

IRISZZW replied:
I will remove it.

ruuddw commented:
Few minor comments, please clarify this is for axs103 and what config.

ruuddw commented on soc/arc/snps_arc_axs/soc.c:
```diff
@@ -0,0 +1,54 @@
+	return 0;
+}
+
+/*SYS_INIT(arc_axs_init, PRE_KERNEL_1, CONFIG_KERNEL_INIT_PRIORITY_DEFAULT);*/
```
Shippable fails on this: 'arc_axs_init' defined but not used [-Werror=unused-function].
If interrupt support is still work in progress, better not include it at all.

IRISZZW replied:
removed

ruuddw reviewed (approved):
I'm ok with merging as (very) "initial" support, but please address the comments raised for next release.

IRISZZW commented:
> I'm ok with merging as (very) "initial" support, but please address the comments raised for next release.

currently, due to the coronavirus, our office is closed. I am working from home with no hardware board. I will update when I get back to office.

abrodkin commented on boards/arc/axs/Kconfig.board:
```diff
@@ -0,0 +1,10 @@
+# DesignWare ARC HS Development Kit board configuration
+
+# Copyright (c) 2019 Synopsys, Inc. All rights reserved.
```
It's February 2020 already :)

abrodkin commented on boards/arc/axs/Kconfig.defconfig:
```diff
@@ -0,0 +1,9 @@
+# Copyright (c) 2019 Synopsys, Inc. All rights reserved.
```
Ditto

abrodkin commented on boards/arc/axs/axs.dts:
```diff
@@ -0,0 +1,29 @@
+/*
+ * Copyright (c) 2019, Synopsys, Inc. All rights reserved.
```
Ditto

abrodkin commented on boards/arc/axs/support/openocd.cfg:
```diff
@@ -0,0 +1,80 @@
+ftdi_layout_init 0x0088 0x008b
+ftdi_channel 1
+
+
```
Please allow specification of Degilent probe by its serial number, see https://github.com/zephyrproject-rtos/zephyr/blob/master/boards/arc/em_starterkit/support/openocd.cfg#L5 and 
 https://github.com/zephyrproject-rtos/zephyr/commit/946ed0e3695a0744f3b58fa741ec0204a8ac8d5f

Basically you need to add here:
```
# Only specify FTDI serial number if it is specified via
# "set _ZEPHYR_BOARD_SERIAL 12345" before reading this script
if { [info exists _ZEPHYR_BOARD_SERIAL] } {
       ftdi_serial $_ZEPHYR_BOARD_SERIAL
}
```

abrodkin commented on boards/arc/axs/board.cmake:
```diff
@@ -0,0 +1,3 @@
+# SPDX-License-Identifier: Apache-2.0
+
```
Please add the following line here:
```
board_runner_args(openocd "--use-elf")
```

See https://github.com/zephyrproject-rtos/zephyr/pull/22891 for more details.

abrodkin commented on soc/arc/snps_arc_axs/soc.h:
```diff
@@ -0,0 +1,44 @@
+
+#include <sys/util.h>
+
+
```
Why do we need all those empty lines?

abrodkin commented on soc/arc/snps_arc_axs/soc.h:
```diff
@@ -0,0 +1,44 @@
+#define IRQ_TIMER0				16
+#define IRQ_TIMER1				17
+
+
```
Ditto

abrodkin commented on soc/arc/snps_arc_axs/soc.h:
```diff
@@ -0,0 +1,44 @@
+#define DT_UART_NS16550_PORT_1_IRQ_FLAGS	0 /* Default */
+#define DT_UART_NS16550_PORT_2_IRQ_FLAGS	0 /* Default */
+
+
```
Ditto

abrodkin commented on dts/arc/arc_axs.dtsi:
```diff
@@ -0,0 +1,75 @@
+			clock-frequency = <33333333>;
+			reg = <0xE0022000 0x1000>;
+			label = "UART_2";
+			interrupts = <19 1>;
```
Just noticed that we have 2 values in `interrupt` property here: `<19 1>`.
Is the second one being really used in Zephyr? Doesn't seem to be the case for me. Care to check?

abrodkin commented on dts/arc/arc_axs.dtsi:
```diff
@@ -0,0 +1,75 @@
+
+		ddr0: memory@80000000 {
+			device_type = "memory";
+			compatible = "mmio-sram";
```
Technically this is not SRAM but true DDR. So I would suggest to drop that bogus compatible.

abrodkin commented on boards/arc/axs/support/openocd.cfg:
```diff
@@ -0,0 +1,80 @@
+set _TARGETNAME1 $_CHIPNAME.cpu1
+jtag newtap $_CHIPNAME cpu1 -irlen 4 -ircapture 0x1 -expected-id 0x200024b1
+
+################################
```
I would suggest to move core 0 definition to the end so that it gets selected by OpenOCD by default, see https://github.com/zephyrproject-rtos/zephyr/commit/7a8c624f53338d52876c120ff1a7afbcd36bd3b9 for more details.

abrodkin commented on boards/arc/axs/support/openocd.cfg:
```diff
@@ -0,0 +1,80 @@
+
+# ARCs supports only JTAG.
+transport select jtag
+
```
Extra newline

abrodkin commented on boards/arc/axs/doc/index.rst:
```diff
@@ -0,0 +1,192 @@
+   :app: <my app>
+   :goals: flash
+
+This command still uses openocd and gdb to load the application elf file to
```
Not any longer. Now `flash` loads `zephyr.elf` with help of OpenOCD and then starts execution from the entry-point. That said GDB is no longer used here at all.

abrodkin reviewed (changes requested):
I made quite some comments, please review them and fix where necessary.
Also I'm not convinced still we should keep using UART w/o interrupts even though hardware setup is very-very similar to HSDK board so nothing should be blocking us here.

IRISZZW committed:
boards: arc: add initial support for ARC axs board.

* add basic support for  GPIO and UART with no
  interrupt, in axs board, IPs in mother board use
  a second-level interrupt which currently not supported.
  currently, we can use the debug uart and gpio buttons,
  leds and seven-segment deisplays.

Signed-off-by: Watson Zeng <zhiwei@synopsys.com>

galak commented:
@ruuddw / @vonhust what's the plan for this board port?

vonhust commented:
drop this PR as AXS board is outdated, no more production,  We will keep it in synopsys's repo for a while, finally discard it.
